Make torch.linalg.eigvalsh differentiable

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57189 Make torch.linalg.eigvalsh differentiable**
* #57188 Make torch.linalg.svdvals differentiable

`torch.linalg.eigvalsh` now supports autograd. This is achieved by
computing the eigenvectors internally if input requires grad,
otherwise the eigenvectors are not computed and the operation is faster.

Differential Revision: [D28199708](https://our.internmc.facebook.com/intern/diff/D28199708)

IvanYashchuk committed:
Make torch.linalg.eigvalsh differentiable

`torch.linalg.eigvalsh` now supports autograd. This is achieved by
computing the eigenvectors internally if input requires grad,
otherwise the eigenvectors are not computed and the operation is faster.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Make torch.linalg.eigvalsh differentiable"


`torch.linalg.eigvalsh` now supports autograd. This is achieved by
computing the eigenvectors internally if input requires grad,
otherwise the eigenvectors are not computed and the operation is faster.

[ghstack-poisoned]

albanD reviewed (approved):
LGTM!

Note that we added faster gradcheck implementation a couple days back (that is the new default in our tests), you might want to benchmark the gradgrad tests again to see if they are now fast enough!

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -453,7 +453,6 @@
 .. seealso::
 
         :func:`torch.linalg.eigvalsh` computes only the eigenvalues.
```
Perhaps add a line explaining that this function does not have unicity / gradient problems?

IvanYashchuk replied:
All gradients in PyTorch should be stable by default, I think we need to document only unstable and unexpected cases.

Lezcano replied:
I just wanted to stress the difference. Perhaps write it as:
> Unlike :func:\`torch.linalg.eigh\`, the gradients of this function are always stable
?

mruberry replied:
Highlighting the difference(s) with the current operation sounds good. Maybe something as simple as "only the eigenvalues, but its gradients are always stable."?

I still don't like referring to gradients as "stable" but that's an issue for another time.

IvanYashchuk replied:
I updated this section and I used the term "numerically stable". Which is a well-defined concept
https://nhigham.com/2020/08/04/what-is-numerical-stability/
https://en.wikipedia.org/wiki/Numerical_stability

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -2049,6 +2049,14 @@ std::tuple<Tensor&, Tensor&> linalg_eigh_out(const Tensor& input, std::string up
+  // the eigenvectors are not exposed to the user
+  if (torch::autograd::compute_requires_grad(input)) {
+    Tensor values;
+    std::tie(values, std::ignore) = at::linalg_eigh(input, uplo);
```
nit. Perhaps it's simpler to just `return std::get<0>(at::linalg_eig(input, uplo))`?

IvanYashchuk replied:
I find the current way more clear: it's easy to understand from the names of the return tensors that we are saving the eigenvalues and ignoring the eigenvectors. `return std::get<0>` is valid, but hides bits of information.

lezcano reviewed (commented):
Short and sweet!

I just left two small comments, but otherwise looks great!

I will create in a bit a PR with the derivative for `linalg.eig`. I'll take the chance to add diff support for `eigvals` there as well following this PR.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -536,9 +535,6 @@
 
-.. note:: This function is not differentiable. If you need differentiability use
-          :func:`torch.linalg.eigh` instead, which also computes the eigenvectors.
-
```
Add a `seealso:` section now, given that there is no explicit reference to `eigh`. E.g.
```
.. seealso::
          :func:`torch.linalg.eig` computes the full eigenvalue decomposition. 
```

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -4151,6 +4156,18 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+           gradcheck_wrapper=gradcheck_wrapper_hermitian_input,
+           decorators=[skipCUDAIfNoMagma, skipCUDAIfRocm, skipCPUIfNoLapack],
+           skips=(
+               # cuda gradchecks are slow
```
Not for this PR, but we should revisit if these are still "slow" after @soulitzer's work on gradcheck performance

What do you think, @soulitzer?

IvanYashchuk replied:
I removed the skip here, see https://github.com/pytorch/pytorch/pull/57189#issuecomment-831773746.
For other OpInfos I'll check later and remove skips in a separate PR if they are not "slow" anymore.

mruberry reviewed (approved):
Cool! @Lezcano's suggestions SGTM, too

IvanYashchuk commented:
Running `pytest test/test_ops.py -k 'eigvalsh' --durations=10` gives
```
========================================================== slowest 10 durations ==========================================================
2.52s call     test/test_ops.py::TestOpInfoCUDA::test_duplicate_method_tests_linalg_eigvalsh_cuda_float32
1.50s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_eigvalsh_cuda_complex64
1.43s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_linalg_eigvalsh_cuda_float32
1.32s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_linalg_eigvalsh_cuda_complex128
1.13s call     test/test_ops.py::TestOpInfoCUDA::test_supported_backward_linalg_eigvalsh_cuda_complex128
1.11s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_linalg_eigvalsh_cpu_complex64
0.84s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_linalg_eigvalsh_cpu_float32
0.76s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_linalg_eigvalsh_cuda_float64
0.52s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_eigvalsh_cuda_complex128
0.48s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_linalg_eigvalsh_cuda_float64

```

The `gradgrad` test is not the slowest one and I removed skipping it.

IvanYashchuk committed:
Update on "Make torch.linalg.eigvalsh differentiable"


`torch.linalg.eigvalsh` now supports autograd. This is achieved by
computing the eigenvectors internally if input requires grad,
otherwise the eigenvectors are not computed and the operation is faster.

[ghstack-poisoned]

lezcano reviewed (approved):
Everything's been addressed on my end.

albanD commented:
On its way to land!

IvanYashchuk committed:
Update on "Make torch.linalg.eigvalsh differentiable"


`torch.linalg.eigvalsh` now supports autograd. This is achieved by
computing the eigenvectors internally if input requires grad,
otherwise the eigenvectors are not computed and the operation is faster.

Differential Revision: [D28199708](https://our.internmc.facebook.com/intern/diff/D28199708)

[ghstack-poisoned]
